Taiko scoreprocessing

smoogipoo opened the pull request:
Prereqs:
- [x] https://github.com/ppy/osu/pull/542

I didn't really want to derive `Hit` once more for `HitFinisher` (which would add no additional properties), and realized it could be used in the future for other hit objects as well, so I added a property to TaikoHitObject.
This'll also be eventually used for actually getting the visual representation of these base hitobjects.

The calculations right now should be equivalent to osu!stable for the purpose of TWC, although we should look at simplifying it uniformly in the future.

smoogipoo committed:
Merge branch 'taiko_drumroll_tick_judgements' into taiko_scoreprocessing

smoogipoo committed:
Make TaikoHitObject store whether the hit object is a finisher.

smoogipoo committed:
Implement Taiko score processing.

smoogipoo committed:
Merge branch 'taiko_drumroll_tick_judgements' into taiko_scoreprocessing

Conflicts:
	osu.Game.Modes.Taiko/TaikoScoreProcessor.cs

smoogipoo committed:
Fix post-merge errors.

smoogipoo committed:
Fix nullref.

smoogipoo committed:
Rename things around to make a bit more sense.

smoogipoo committed:
Move ScoreProcessor and Score to Scoring/.

peppy commented on osu.Game.Modes.Taiko/Scoring/TaikoScoreProcessor.cs:
```diff
@@ -0,0 +1,270 @@
+        /// This is legacy and should be fixed, but is kept as is for now for compatibility.
+        /// </para>
+        /// </summary>
+        private const double hp_hit_good_max = hp_hit_good * 8;
```
isn't this magnitudes worse since you're not considering hp drain variable?

peppy replied:
okay i see these are just base values which are interpolated later. you need to explain what these variables are better (nowhere is the `_max` explained).

also i'd just fix this rather than keeping broken "compatibility"

smoogipooo replied:
Fix it to what? Ideally the entire HP system would be reworked, but as the plan is to use this for TWC then I want to maintain as much compatibility as possible (at least for until TWC is over).
Otherwise someone fails on screen but they haven't failed in-game and the entire audience notices the inconsistency.

peppy replied:
are there any tournament level maps that use hp drain <4?

peppy replied:
as for what to fix it to: choose a sane constant that makes sense to you :p

smoogipooo replied:
There is one map with hp = 4 in the quarterfinals. I'll make the change pre-emptively, but will be ready to change things back if that changes in the finals.

peppy commented on osu.Game.Modes.Taiko/Beatmaps/TaikoBeatmapConverter.cs:
```diff
@@ -43,12 +44,15 @@ private TaikoHitObject convertHitObject(HitObject original)
             IHasRepeats repeatsData = original as IHasRepeats;
             IHasEndTime endTimeData = original as IHasEndTime;
 
+            bool isFinisher = ((original.Sample?.Type ?? SampleType.None) & SampleType.Finish) > 0;
```
can we use a better term than "finisher"? maybe `IsLargeHit`

peppy replied:
or just `Large` as the field

peppy commented on osu.Game.Modes.Taiko/Scoring/TaikoScoreProcessor.cs:
```diff
@@ -0,0 +1,270 @@
+        private double accuracyScore => accuracy_portion_max * Math.Pow(Accuracy, 3.6) * totalHits / maxTotalHits;
+
+        /// <summary>
+        /// The final bonus score.
```
the cumulative bonus score? it's only final if the play is complete.

peppy replied:
same goes for other uses of "final" above, doesn't feel correct

peppy commented on osu.Game.Modes.Taiko/Scoring/TaikoScoreProcessor.cs:
```diff
@@ -0,0 +1,270 @@
+
+        protected override void UpdateCalculations(TaikoJudgement newJudgement)
+        {
+            var tickJudgement = newJudgement as TaikoDrumRollTickJudgement;
```
you never actually use this casted version. consider making this `bool isTick = newJudgement is ...`

peppy commented on osu.Game.Modes.Taiko/Scoring/TaikoScoreProcessor.cs:
```diff
@@ -0,0 +1,270 @@
+            maxComboPortion = comboPortion;
+        }
+
+        protected override void UpdateCalculations(TaikoJudgement newJudgement)
```
newJudgement -> judgement or j

peppy replied:
in fact, maybe this method should be renamed... it's more like `OnNewJudgement`

peppy commented on osu.Game.Modes.Taiko/Scoring/TaikoScoreProcessor.cs:
```diff
@@ -0,0 +1,270 @@
+            var tickJudgement = newJudgement as TaikoDrumRollTickJudgement;
+
+            // Don't consider ticks as a type of hit that counts towards map completion
+            if (tickJudgement == null)
```
i'd also restructure all the code below to be

```
if (isTick)
{
    //do everything tick related
}
else
{
    //everything else
}
```

i think it will simplify things. you may even be able to do an early return for ticks, remove the else{} and reduce nesting

smoogipoo committed:
Rename finisher -> accented.

smoogipoo committed:
Final -> cumulative, to represent a growing score.

smoogipoo committed:
Use isTick.

smoogipoo committed:
More accented renaming + improve comments.

smoogipoo committed:
Remove linear interpolation of hp_hit_good.

I think this is an okay compromise because hp_hit_great doesn't use linear interpolation here. I am tentative on this change, but I need to code up the rest of taiko before I can test this further.

smoogipoo committed:
Rename UpdateJudgement -> OnNewJugement + rename judgement parameter.

peppy committed:
Merge branch 'master' into taiko_scoreprocessing
